python/ray/autoscaler/command_runner.py

```diff
@@ -327,6 +344,286 @@ def _set_ssh_ip_if_required(self):
+            # output processing will get overwhelemed with interactive output
+            # e.g. `pip install` outputs HUNDREDS of progress-bar lines
+            # and we have to read + write all of them with non-trival overhead
+            #
+            # after all, even just printing output to console can often slow
+            # down a fast-printing app, and we do more than just print, and
+            # all that from Python, which is much slower than C in stream
+            # processing
+            if stdout_file is None:
+                stdout_file = self.process_runner.DEVNULL
+            if stderr_file is None:
+                stderr_file = self.process_runner.DEVNULL
+
+            return self.process_runner.check_call(
+                cmd,
+                # Do not inherit stdin as it messes with bash signals
+                # (ctrl-C for SIGINT) and these commands aren't supposed to
+                # take input anyway.
+                stdin=self.process_runner.PIPE,
+                stdout=stdout_file,
+                stderr=stderr_file)
+
+        with self.process_runner.Popen(
+                cmd,
+                # Do not inherit stdin as it messes with bash signals
+                # (ctrl-C for SIGINT) and these commands aren't supposed to
+                # take input anyway.
+                stdin=self.process_runner.PIPE,
+                stdout=self.process_runner.PIPE,
+                stderr=self.process_runner.PIPE,
+                bufsize=1,  # line buffering
+                universal_newlines=True  # text mode outputs
+        ) as p:
+            from concurrent.futures import ThreadPoolExecutor
+
+            # Closing stdin might be necessary to signal EOF to some
+            # apps (they might get stuck waiting for input forever otherwise).
+            p.stdin.close()
+
+            # We do NOT USE the pooling feature of `ThreadPoolExecutor`.
+            # We ONLY rely on it's ability to create futures from threads.
+            # This instance will be `shutdown()` ASAP so it's fine to
+            # create one in such a weird place.
+            #
+            # This approach allows us to HAVE NO CUSTOM SYNCHRONIZATION by
+            # off-loading the return value and exception passing to the
+            # standard library (`ThreadPoolExecutor` internals).
+            #
+            # The code is thus 100% thread-safe as long as the stream readers
+            # are read-only except for return values and possible exceptions.
+            with ThreadPoolExecutor(max_workers=2) as pool:
+                stdout_future = \
+                    pool.submit(self._read_subprocess_stream,
+                                p.stdout, stdout_file,
+                                is_stdout=True)
+                stderr_future = \
+                    pool.submit(self._read_subprocess_stream,
+                                p.stderr, stderr_file,
+                                is_stdout=False)
```

Comment:
tip: you should just rename this. 

```suggestion
            with ThreadPoolExecutor(max_workers=2) as executor:
                stdout_future = \
                    executor.submit(self._read_subprocess_stream,
                                p.stdout, stdout_file,
                                is_stdout=True)
                stderr_future = \
                    executor.submit(self._read_subprocess_stream,
                                p.stderr, stderr_file,
                                is_stdout=False)
```